sync/statesync/code_syncer.go

```diff
@@ -165,76 +159,22 @@ func (c *codeSyncer) work(ctx context.Context) error {
 // marks the work as complete.
 // codeHashes should not be empty or contain duplicate hashes.
 // Returns an error if one is encountered, signaling the worker thread to terminate.
-func (c *codeSyncer) fulfillCodeRequest(ctx context.Context, codeHashes []common.Hash) error {
+func (c *CodeSyncer) fulfillCodeRequest(ctx context.Context, codeHashes []common.Hash) error {
 	codeByteSlices, err := c.client.GetCode(ctx, codeHashes)
 	if err != nil {
 		return err
 	}
 
-	// Hold the lock while modifying outstandingCodeHashes.
-	c.lock.Lock()
 	batch := c.db.NewBatch()
 	for i, codeHash := range codeHashes {
 		customrawdb.DeleteCodeToFetch(batch, codeHash)
-		c.outstandingCodeHashes.Remove(codeHash)
 		rawdb.WriteCode(batch, codeHash, codeByteSlices[i])
+		// Mark not in-flight after persisting.
+		c.inFlight.Delete(codeHash)
```

Comment:
This needs to come after `batch.Write()` as if it returns error then another worker could see that the hash isn't in-flight, and then its database check races with the `Write()` so it duplicates the work.